File: VB/AuditDemo.Module/MSSqlServerTimestampStrategy.vb
Imports System
Imports DevExpress.Xpo
Imports DevExpress.Persistent.AuditTrail

Namespace AuditDemo.Module
	Public Class MSSqlServerTimestampStrategy
		Implements IAuditTimestampStrategy

		Private cachedTimeStamp As DateTime
		#Region "IAuditTimeStampStrategy Members"
		Public Function GetTimestamp(ByVal auditDataItem As AuditDataItem) As DateTime
			Return cachedTimeStamp
		End Function
		Public Sub OnBeginSaveTransaction(ByVal session As Session)
			cachedTimeStamp = DirectCast(session.ExecuteScalar("select getdate()"), DateTime)
		End Sub
		#End Region
	End Class
End Namespace

File: CS/AuditDemo.Module/MSSqlServerTimestampStrategy.cs
using System;
using DevExpress.Xpo;
using DevExpress.Persistent.AuditTrail;

namespace AuditDemo.Module {
	public class MSSqlServerTimestampStrategy : IAuditTimestampStrategy{
		DateTime cachedTimeStamp;
		#region IAuditTimeStampStrategy Members
		public DateTime GetTimestamp(AuditDataItem auditDataItem) {
			return cachedTimeStamp;
		}
		public void OnBeginSaveTransaction(Session session) {
            cachedTimeStamp = (DateTime)session.ExecuteScalar("select getdate()");
		}
		#endregion
	}
}
